### PR TITLE
430: Source deduction at Order complete

### PR DESCRIPTION
**Pull request by seruymt**
### Fixed Issues (if relevant)
1. https://github.com/magento-engcom/msi/issues/430

This is one of the ways how it can be implemented

### CODE REVIEW

File: app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            return $this;
+        }
+
+        $order = $shipment->getOrder();
+
+        // I'm not sure about websiteId
+        $websiteId = $order->getStore()->getWebsiteId();
+
+        foreach ($shipment->getItems() as $item) {
+            if (!$item->getSources()) {
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php]: there are no get/setSource methods in the 
`\Magento\Sales\Api\Data\ShipmentItemInterface`

File: app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                continue;
+            }
+            foreach ($item->getSources() as $source) {
+                $source_code = $source['source_code'];
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php]: use CamelCase for variable names, like 
`sourceCode `

File: app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            ->setSku($sku)
+            ->setQuantity($qty)
+            ->setStockId($stockId)
+            ->build();
+        $this->appendReservations->execute([$reservation]);
+    }
+
+    /**
+     * Get Source item by source_code and product SKU
+     * TODO: Maybe need to implement separate method as API
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php]: yes, makes sense

File: app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        foreach ($shipment->getItems() as $item) {
+            if (!$item->getSources()) {
+                continue;
+            }
+            foreach ($item->getSources() as $source) {
+                $source_code = $source['source_code'];
+                $qty = (float)$source['qty'];
+                $sourceItem = $this->getSourceItemByCodeAndSku($source_code, $item->getSku());
+                if (($sourceItem->getQuantity() - $qty) >= 0) {
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php]: ideally is to have extension point here to allow making a deduction for Backorders, thus 
we will do 
`$sourceItem->setQuantity($sourceItem->getQuantity() - $qty);` even when

`($sourceItem->getQuantity() - $qty) < 0`

File: app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                    $sourceItem->setQuantity($sourceItem->getQuantity() - $qty);
+                    $this->sourceItemsSave->execute([$sourceItem]);
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php]: why do we make SourceItem persistence at once, but don't collect all source items need to be persisted.
And save all of them at once. That should be more efficient

File: app/code/Magento/InventorySales/Plugin/Sales/Shipment/CollectSourcesForShipmentItems.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        foreach ($shipment->getItems() as $item) {
+            if (!empty($sources[$item->getOrderItemId()])) {
+                $sourcesToDeduct = [];
+                foreach ($sources[$item->getOrderItemId()] as $data) {
+                    $sourcesToDeduct[] = [
+                        'source_code' => $data->getSourceCode(),
+                        'qty' => $data->getQtyToDeduct()
+                    ];
+                }
+                $item->setSources($sourcesToDeduct);
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Plugin/Sales/Shipment/CollectSourcesForShipmentItems.php]: there is no `setSources` method in `Magento\Sales\Api\Data\ShipmentItemInterface` interface

Extension Attribute mechanism could be used here

File: app/code/Magento/InventorySales/Plugin/Sales/Shipment/CollectSourcesForShipmentItems.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        //TODO: Temporary decision. Need to implement logic with UI part (get data from items array)
+        $sources = $this->getDataFromShippingAlgorithm($order);
+
+        foreach ($shipment->getItems() as $item) {
+            if (!empty($sources[$item->getOrderItemId()])) {
+                $sourcesToDeduct = [];
+                foreach ($sources[$item->getOrderItemId()] as $data) {
+                    $sourcesToDeduct[] = [
+                        'source_code' => $data->getSourceCode(),
+                        'qty' => $data->getQtyToDeduct()
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Plugin/Sales/Shipment/CollectSourcesForShipmentItems.php]: let's rename qty - > qtyToDeduct to make it more readable

File: app/code/Magento/InventorySales/Plugin/Sales/Shipment/CollectSourcesForShipmentItems.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * @param ShipmentFactory $subject
+     * @param callable $proceed
+     * @param Order $order
+     * @param array $items
+     * @param null $tracks
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundCreate(
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Plugin/Sales/Shipment/CollectSourcesForShipmentItems.php]: why do you need `aroundCreate ` but not `afterCreate` ? 
as you call process as the first line of your function
`$shipment = $proceed($order, $items, $tracks);`

**Review by maghamed — CHANGES_REQUESTED** [review]
Currently in one of integration tests we have a situation when SKU is not set for OrderItem
```
            $itemSku = $orderItem->getSku();
```
https://travis-ci.org/magento-engcom/msi/jobs/346173533

**Comment by smoskaluk** [comment]
@seruymt For Source deduction it should be taken into account if source is disabled or stock status for assigned source to the product is "Out of stock"

File: app/code/Magento/InventorySourceSelectionApi/Api/SourceSelectionServiceInterface.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+interface SourceSelectionServiceInterface
+{
+    /**
+     * @param InventoryRequestInterface $inventoryRequest
+     * @param string $algorithmCode
+     * @return SourceSelectionResultInterface
+     */
+    public function execute(
+        InventoryRequestInterface $inventoryRequest,
+        string $algorithmCode = 'priority'
```
Comment (maghamed) [inline app/code/Magento/InventorySourceSelectionApi/Api/SourceSelectionServiceInterface.php]: it's better not to introduce interfaces with a default parameter. 
Default parameters are in most cases are evil.

My point is that whether you use default arguments, it’s extremely unlikely that you’re ever going to be able to change default values once you’ve published an API. 

To gain a maximum flexibility - it's better to introduce dedicated Service.
`GetDefaultSourceSelectionAlgorithm` which would be used in client's code.

**Review by maghamed — APPROVED** [review]

**Pull request merged**